### PR TITLE
Issue 910  matvec op

### PR DESCRIPTION
**Pull request by kohr-h**
Fix for #910 and move of `MatVecOperator` from `space/npy_ntuples` to `operator/tensor_ops`, plus rename from `MatVecOperator` to `MatrixOperator`.
No other work like docs or test improvement since #861 also has its part in it.

### CODE REVIEW

File: odl/operator/tensor_ops.py [hunk 1]
```diff
@@ -692,6 +696,141 @@ def __init__(self, vfspace, weighting=None):
         super().__init__(vfspace, vecfield=ones, weighting=weighting)
 
 
+class MatrixOperator(Operator):
+
+    """Matrix multiply operator :math:`\mathbb{F}^n -> \mathbb{F}^m`.
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: Matrix-vector multiplication operator.

Also move math down a line

File: odl/operator/tensor_ops.py [hunk 2]
```diff
@@ -692,6 +696,141 @@ def __init__(self, vfspace, weighting=None):
...
+
+    def __init__(self, matrix, domain=None, range=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        matrix : `array-like` or  `scipy.sparse.spmatrix`
+            Matrix representing the linear operator. Its shape must be
+            ``(m, n)``, where ``n`` is the size of ``domain`` and ``m`` the
+            size of ``range``. Its dtype must be castable to the range
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: Either typeset `dtype` or dont.

File: odl/operator/tensor_ops.py [hunk 3]
```diff
@@ -692,6 +696,141 @@ def __init__(self, vfspace, weighting=None):
...
+        else:
+            self.__matrix = np.asarray(matrix)
+
+        if self.matrix.ndim != 2:
+            raise ValueError('matrix {} has {} axes instead of 2'
+                             ''.format(matrix, self.matrix.ndim))
+
+        # Infer domain and range from matrix if necessary
+        if domain is None:
+            domain = NumpyFn(self.matrix.shape[1], dtype=self.matrix.dtype)
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: `domain = fn(self.matrix.shape[1], dtype=self.matrix.dtype)`

File: odl/operator/tensor_ops.py [hunk 4]
```diff
@@ -692,6 +696,141 @@ def __init__(self, vfspace, weighting=None):
...
+        elif not isinstance(domain, FnBase):
+            raise TypeError('`domain` {!r} is not an `FnBase` instance'
+                            ''.format(domain))
+
+        if range is None:
+            range = NumpyFn(self.matrix.shape[0], dtype=self.matrix.dtype)
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: `range = fn(self.matrix.shape[0], dtype=self.matrix.dtype)`

File: odl/operator/tensor_ops.py [hunk 5]
```diff
@@ -692,6 +696,141 @@ def __init__(self, vfspace, weighting=None):
...
+                            'range data type {!r}'
+                            ''.format(domain.dtype, range.dtype))
+
+        if self.matrix.shape != (range.size, domain.size):
+            raise ValueError('matrix shape {} does not match the required '
+                             'shape {} of a matrix {} --> {}'
+                             ''.format(self.matrix.shape,
+                                       (range.size, domain.size),
+                                       domain, range))
+        if not np.can_cast(self.matrix.dtype, range.dtype):
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: weird order of the dtype checks, shape check shouldnt be interlaced

**Comment by kohr-h** [comment]
Comments fixed, doctest and repr added. Merging after CI

File: odl/operator/tensor_ops.py [hunk 6]
```diff
@@ -692,6 +696,210 @@ def __init__(self, vfspace, weighting=None):
...
+    """
+
+    def __init__(self, matrix, domain=None, range=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        matrix : `array-like` or `scipy.sparse.base.spmatrix`
+            2-dimensional array representing the linear operator.
+            For Scipy sparse matrices only rank-1 tensor spaces are
```
Comment (kohr-h) [inline odl/operator/tensor_ops.py]: need to change this

File: odl/operator/tensor_ops.py [hunk 7]
```diff
@@ -692,6 +696,208 @@ def __init__(self, vfspace, weighting=None):
...
+        >>> dom = odl.uniform_discr(0, 1, 4)
+        >>> ran = odl.uniform_discr(0, 1, 3)
+        >>> op = MatrixOperator(m, domain=dom, range=ran)
+        >>> op(dom.one())
+        uniform_discr(0.0, 1.0, 3).element([4.0, 4.0, 4.0])
+
+        Notes
+        -----
+        For a matrix :math:`A \\in \\mathbb{F}^{n \\times m}`, the
+        operation on a tensor :math:`T \\in \mathbb{F}^{n_1 \\times
```
Comment (kohr-h) [inline odl/operator/tensor_ops.py]: screw it, this needs to be changed, too

**Comment by kohr-h** [comment]
Fixed repr for sparse matrix and added a doctest for that case, too. Now ready for merge after CI

File: odl/operator/tensor_ops.py [hunk 8]
```diff
@@ -692,6 +696,215 @@ def __init__(self, vfspace, weighting=None):
...
+            # Don't convert to dense, can take forever
+            matrix_str = repr(self.matrix)
+        else:
+            matrix_str = np.array2string(self.matrix, separator=', ')
+        posargs = [matrix_str]
+
+        # Optional arguments with defaults, inferred from the matrix
+        optargs = []
+        # domain
+        optargs.append(
```
Comment (adler-j) [inline odl/operator/tensor_ops.py]: extremely verbose way of writing

```
optargs = [('domain', self.domain, fn(self.matrix.shape[1],
                                      self.matrix.dtype)),
           ('range', self.range, fn(self.matrix.shape[0],
                                    self.matrix.dtype))]
```

**Comment by kohr-h** [comment]
Okay, I'll fix that stuff on #861.

**Comment by adler-j** [comment]
Good!

**Pull request merged**